File: src/services/instructions/alu/ALU_W_REGX_REGY.java
package emulator.source.alu

import emulator.engine.CpuContext
 import Instruction from '../Instruction'import REGISTER_VALUE_NAME_MAPPER from '@/constants/registers'

export default class ALU_W_REGX_REGY extends Instruction {
	int type
	
	public ALU_W_REGX_REGY(short[] memory, int address, int source, int destination, int type) {
		super(memory, address, source, destination)
		super.setAssembler(Instruction.getTypeStr(type) + this.sdest + ", " + this.ssource)
		this.type = type
	}

	
	exec ({ context, memory }) {
		int old_a = context.getReg(this.destination) 
		long res = 0
		switch (type) {
		case ADD_W: res = context.getReg(this.destination)  + context.getReg(this.source)  break
		case SUB_W: res = context.getReg(this.destination)  - context.getReg(this.source)  break
		case AND_W: res = context.getReg(this.destination)  & context.getReg(this.source)  break
		case OR_W : res = context.getReg(this.destination)  | context.getReg(this.source)  break
		case XOR_W: res = context.getReg(this.destination)  ^ context.getReg(this.source)  break
		case SHL_W: res = context.getReg(this.destination)  << context.getReg(this.source)  break
		case SHR_W: res = context.getReg(this.destination)  >>> context.getReg(this.source)  break
		case MUL_W:	res = context.getReg(this.destination)  * context.getReg(this.source)  
					context.h  = (short)((res & 0xffff0000) >> 16)
					break
		case DIV_W: 	res = context.getReg(this.destination)  / context.getReg(this.source)  
					context.h  = (short)(context.getReg(this.destination)  % context.getReg(this.source) )
					break
		default: throw new RuntimeException("Unsupported operation type: " + type)
		}
		context.getReg(this.destination)  = (int)res
		markFlags(res, context.getReg(this.destination) , context)
		markOverflow(old_a, context.getReg(this.source) , context.getReg(this.destination) , context)
		context.pc  += 2
	}
}

File: src/services/instructions/alu/AluWRegXRegY.js
import Instruction from '../Instruction'
import { REGISTER_VALUE_NAME_MAPPER } from '@/constants/registers'
import { INSTRUCTIONS_TYPES_NAMES, INSTRUCTION_TYPES } from '@/constants/instructions'
import { INSTRUCTIONS_TYPE_FUNCTION_COMPUTATIONS } from '@/helpers/instruction'

const ASSEMBLER_INSTRUCTION_EXPRESSION = (type, sdestination, ssource) => `${type} ${sdestination}, ${ssource}`

export default class AluWRegXRegY extends Instruction {
	constructor (memory, address, source, destination, type, symbolTable) {
		super(memory, address, source, destination, symbolTable)
		super.setAssembler(
			ASSEMBLER_INSTRUCTION_EXPRESSION(
				INSTRUCTIONS_TYPES_NAMES[type],
				this.sdestination,
				this.ssource
			))

		this.type = type
	}


	exec ({ context }) {
		const old_a = context[REGISTER_VALUE_NAME_MAPPER[this.destination]]

		// TODO: type `long` -->> long result = 0
		let result = 0

		result = INSTRUCTIONS_TYPE_FUNCTION_COMPUTATIONS[this.type](context[REGISTER_VALUE_NAME_MAPPER[this.destination]], context[REGISTER_VALUE_NAME_MAPPER[this.source]])

		switch (this.type) {
			case INSTRUCTION_TYPES.MUL_W:
				//TODO: short type cast -->> context.h  = (short)((result & 0xffff0000) >> 16)
				context.h  = (result & 0xffff0000) >> 16
				break
		  case INSTRUCTION_TYPES.DIV_W:
				//TODO: short type cast -->> context.h  = (short)(context[REGISTER_VALUE_NAME_MAPPER[this.destination]]  % context[REGISTER_VALUE_NAME_MAPPER[this.source]] )
				context.h  = context[REGISTER_VALUE_NAME_MAPPER[this.destination]]  % context[REGISTER_VALUE_NAME_MAPPER[this.source]]
				break
		}

		// TODO: int cast -->> context[REGISTER_VALUE_NAME_MAPPER[this.destination]]  = (int)result
		context[REGISTER_VALUE_NAME_MAPPER[this.destination]] = result
		Instruction.markFlags(result, context[REGISTER_VALUE_NAME_MAPPER[this.destination]] , context)
		Instruction.markOverflow(old_a, context[REGISTER_VALUE_NAME_MAPPER[this.source]] , context[REGISTER_VALUE_NAME_MAPPER[this.destination]] , context)
		context.pc  += 2
	}
}
